nemo/collections/asr/parts/k2/graph_transducer.py

```diff
@@ -0,0 +1,480 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from contextlib import nullcontext
+from typing import ContextManager
+import torch
+import torch.nn.functional as F
+
+from nemo.core.classes.loss import Loss
+from nemo.core.utils.k2_guard import k2
+
+
+def force_float32_context() -> ContextManager:
+    """Get context manager to force float32 precision in autocast mode."""
+    if torch.is_autocast_enabled():
+        return torch.cuda.amp.autocast(dtype=torch.float32)
+    return nullcontext()
+
+
+class GraphTransducerLossBase(Loss):
+    """
+    Base class for graph transducer losses.
+    Implementation of the approach described in "Powerful and Extensible WFST Framework for RNN-Transducer Losses"
+    https://ieeexplore.ieee.org/document/10096679
+
+    Compose-Transducer: compose the unit (target text) and temporal schemas (graphs) into lattice.
+        Subclass should implement `get_unit_schema` and `get_temporal_schema` methods.
+    Grid-Transducer: construct the RNN-T lattice (grid) directly in code.
+        Subclass should implement `get_grid` method.
+    """
+
+    def __init__(
+        self, use_grid_implementation: bool, connect_composed=False, double_scores=False, cast_to_float32=False
+    ):
+        """
+
+        Args:
+            use_grid_implementation: Whether to use the grid implementation (Grid-Transducer).
+            connect_composed: Connect graph after composing unit and temporal schemas (only for Compose-Transducer).
+                `connect` operation is slow, it is useful for visualization, but not necessary for loss computation.
+            double_scores: Use calculation of loss in double precision (float64) in the lattice.
+                Does not significantly affect memory usage since the lattice is ~V/2 times smaller
+                than the joint tensor.
+            cast_to_float32: Force cast joint tensor to float32 before log-softmax calculation.
+        """
+        super().__init__()
+        self.use_grid_implementation = use_grid_implementation
+        self.connect_composed = connect_composed
+        self.double_scores = double_scores
+        self.cast_to_float32 = cast_to_float32
+
+    @abc.abstractmethod
+    def get_unit_schema(self, units_tensor: torch.Tensor, vocab_size: int) -> "k2.Fsa":
+        """
+        Get unit schema (target text) graph for Compose-Transducer.
+
+        Args:
+            units_tensor: tensor with target text
+            vocab_size: number of labels (including blank). Needed to construct additional eps-arcs (in some cases).
+
+        Returns:
+            unit schema graph (k2.Fsa).
+            Labels: <unit>:<unit>:<unit_position> (k2.Fsa: labels, aux_labels, unit_positions)
+        """
+        pass
```

Comment:
Maybe raise `NotImplementedError`?

Reply:
`NotImplementedError` is redundant in abstract method